### PR TITLE
Fix stack overflow in stream converters

### PR DESCRIPTION
**Pull request by johanandren**
## References
Refs #24304

## Changes
Replaces accumulating lambdas until `finish` is called (lazy) with eager accumulation to avoid getting a stack overflow when the stream has many elements.

To safely deal with having mutable elements in the stream across multiple materializations instanciating the collector is deferred by having a two step abstraction where the first step delays calling the factory until the first element arrives, that guarantees it will happen once per materialization and not share the mutable accumulator.

The same needed to be done for both the `CollectorState` and the `ReducerState`.

I'm not sure about the rationale for making the state lazy and building a lambda-per-element chain that is evaluated in the end to start with, this change trades the laziness for safety and probably performance (less GC work if nothing else).

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/StreamConvertersSpec.scala [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+import akka.stream.ActorMaterializer
+import akka.stream.testkit.StreamSpec
+import akka.stream.testkit.Utils.TE
+import akka.testkit.DefaultTimeout
+import org.scalatest.time.{ Millis, Span }
+
+import scala.concurrent.Await
+import scala.concurrent.duration._
+
+class StreamConvertersSpec extends StreamSpec with DefaultTimeout {
```
Comment (raboof) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/StreamConvertersSpec.scala]: Great to pull these into their own file :+1:

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/StreamConvertersSpec.scala [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+        override def close(): Unit = closed = true
+      }
+
+      Await.ready(StreamConverters.fromJavaStream(() => new FailingStream[Unit]).runWith(Sink.ignore), 3.seconds)
+
+      closed should ===(true)
+    }
+
+    // Repeater for #24304
```
Comment (raboof) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/StreamConvertersSpec.scala]: :+1:

**Review by raboof — APPROVED** [review]
Pretty explicit, but that makes it clear what's going on so :+1: 

(insert content advisory badge here)

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala [hunk 3]
```diff
@@ -111,11 +111,14 @@ object StreamConverters {
    */
   def javaCollector[T, R](collectorFactory: () => java.util.stream.Collector[T, _ <: Any, R]): Sink[T, Future[R]] =
     Flow[T]
-      .fold(() => new CollectorState[T, R](collectorFactory().asInstanceOf[Collector[T, Any, R]])) {
-        (state, elem) => () =>
-          state().update(elem)
+      .fold {
+        new FirstCollectorState[T, R](collectorFactory.asInstanceOf[() => java.util.stream.Collector[T, Any, R]]): CollectorState[
+          T,
```
Comment (chbatey) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: funky formatting

**Review by chbatey — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala (hunk 3 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala]: The funky formatter stubbornly refuses to let me format it in a sane way.

**Pull request merged**